crates/ruff_python_parser/src/lexer.rs

```diff
@@ -428,10 +429,46 @@ impl<'source> Lexer<'source> {
                     self.cursor.bump();
                     value.push('\\');
                 }
+                '?' => {
+                    self.cursor.bump();
+                    let mut question_count: u8 = 1;
+                    while self.cursor.eat_char('?') {
+                        question_count += 1;
+                    }
+
+                    // The help end magic command is valid if it has 1 or 2 question marks.
+                    // If the `value` is empty i.e., we are at the start of the line,
+                    // then we can't be sure if it is a help end magic command or not.
+                    if question_count > 2
+                        || previous_char_is_whitespace
+                        || value.is_empty()
+                        || !matches!(self.cursor.first(), '\n' | '\r' | EOF_CHAR)
+                    {
+                        // Not a help end magic command, so continue with the lexing.
+                        value.reserve(question_count as usize);
+                        for _ in 0..question_count {
+                            value.push('?');
+                        }
+                        continue;
+                    }
+
+                    if kind.is_help() {
+                        value = value.trim_start_matches([' ', '?']).to_string();
+                    } else {
+                        value.insert_str(0, &kind.to_string());
+                    }
+                    let kind = match question_count {
+                        1 => MagicKind::Help,
+                        2 => MagicKind::Help2,
+                        _ => unreachable!("`question_count` is always 1 or 2"),
+                    };
+                    return Tok::MagicCommand { kind, value };
+                }
                 '\n' | '\r' | EOF_CHAR => {
                     return Tok::MagicCommand { kind, value };
                 }
                 c => {
+                    previous_char_is_whitespace = c.is_whitespace();
```

Comment:
Nit: We should use `is_python_whitespace` here. Python only allows a subset of the unicode whitespace characters.